asab/library/providers/git.py

```diff
@@ -0,0 +1,127 @@
+import tempfile
+import logging
+import os
+import shutil
+
+import pygit2
+
+from .filesystem import FileSystemLibraryProvider
+from ...config import Config
+
+#
+
+L = logging.getLogger(__name__)
+
+#
+
+
+class GitLibraryProvider(FileSystemLibraryProvider):
+	"""
+	Read-only git provider to read from remote repository.
+	It clones a remote git repository to a temporary directory and then uses the
+	FileSystemLibraryProvider to read the files.
+	To read from local git repository, please use FileSystemProvider.
+
+	Configuration:
+	(Use either deploytoken, publickey+privatekey for SSH option, or username and password and HTTP access.)
+	```
+	[library]
+	providers=git+<URL or deploy token>
+
+	[library.GitProvider]
+	publickey=<absolute path to file>
+	privatekey=<absolute path to file>
+	username=johnsmith
+	password=secretpassword
+	```
+	"""
+	def __init__(self, library, path):
+		self.URL = path[4:]
+		self.Callbacks = pygit2.RemoteCallbacks(get_git_credentials(self.URL))
+		self.RepoPath = tempfile.TemporaryDirectory().name
+		self.GitRepository = clone_repository(self.URL, self.RepoPath, self.Callbacks)
+		self._check_remote()
+		super().__init__(library, self.RepoPath)
+
+		from ...proactor import Module
+		self.App.add_module(Module)
+		self.ProactorService = self.App.get_service("asab.ProactorService")
+
+		self.App.PubSub.subscribe("Application.tick/60!", self._periodic_pull)
+
+
+	async def finalize(self, app):
+		clean_path(self.RepoPath)
+
+	async def pull(self):
+		"""
+		Equivalent to `git pull` command.
+		"""
+		await self.ProactorService.execute(pull, self.GitRepository, self.Callbacks)
+
+	def _periodic_pull(self, event_name):
+		self.pull()
```

Comment:
You are calling async code from sync def ...

Reply:
:see_no_evil: shame on me